### ISSUE TITLE
can't understand the flink inference example Resnet50ImageClassification

### ISSUE THREAD

**Issue by linjiaqin** [intel-analytics/analytics-zoo#1715]
[https://github.com/intel-analytics/analytics-zoo/blob/master/apps/model-inference-examples/model-inference-flink/src/main/scala/com/intel/analytics/zoo/apps/model/inference/flink/Resnet50ImageClassification/ImageClassificationStreaming.scala](url)

 I have three questions hoped to be solved.
Fisrt, from the name of the function, I thought the function "matToNCHWAndArray" is to
change the BGR,HWC opencvmat to BGR,CWH.
But we can see it wiil be given to a Jtensor of shape (1, 224,224,3) in the ImageClassificationStreaming.scala. I can't understand. Why not (1, 3, 244, 244)?
Dose it mean the function "matToNCHWAndArray" didn't change the opencvmat to CHWArray?

Second, what's the preprocess for images of the Resnet50 model in the zoo? 
Is it just load the images ,resize to (244,244), change into np.float?
not normalize? not divided by mean?

Third, why should I provide "inputShape: Array[Int], ifReverseInputChannels: Boolean, meanValues: Array[Float], scale: Float" such parameters  for the function "loadtf()"?
Does it will do such operation reverse the channels, divide the meanValus for the images I send to predict?  

Here are some main code I can't understand from the example:
```
def matToNCHWAndArray(mat: OpenCVMat) = {
    val (height, width, channel) = (mat.height(), mat.width(), mat.channels())
    val data = new Array[Float](height * width * channel)
    OpenCVMat.toFloatPixels(mat, data)
    val resArray = new Array[Float](height * width * channel)
    for (h <- 0 to height - 1) {
      for (w <- 0 to width - 1) {
        for (c <- 0 to 2) {
          resArray(c * 224 * 224 + h * 224 + w) = data(h * 224 * 3 + w * 3 + c)
        }
      }
    }
    resArray
  }
```

ImageClassificationStreaming.scala

```
val inputs = fileList.map(file => {
      val imageBytes = FileUtils.readFileToByteArray(file)
      val imageProcess = new ImageProcessor
      val res = imageProcess.preProcess(imageBytes, 224, 224, 123, 116, 103, 1.0)
      val input = new JTensor(res, Array(1, 224, 224, 3))
      List(util.Arrays.asList(input)).asJava
    })
```

**Comment by qiyuangong** [comment]
Hi @linjiaqin 

Your questions are mainly related to image reprocessing and OpenVINO. Let's go through your questions one by one:

1. NHWC and NCHW problem
**When converting pre-trained models, OpenVINO will change input layout to NCHW by default. Because NCHW performs better on CPU.** You can read more from this page (https://intel.github.io/mkl-dnn/understanding_memory_formats.html). Actually, this setting can be changed, but we doesn't expose this parameter in analytics-zoo, so input should be NCHW, We will address this issue in near future.

2. normalize and mean related to OpenVINO
**OpenVINO will include normalization and mean, if means and scale are passed into it.** That increases overall performance on CPU. In our analytics-zoo API, we passed means and scale by default.

3. ifReverseInputChannels
This parameter is inherited from OpenVINO. It is used to reverse input change (BGR->RGB). Because OpenCV uses BGR, while most models use RGB. They requires to reverse iamge changes before feeding into model.

For more information about OpenVINO parameters, pls refer to [Converting a TensorFlow* Model](https://docs.openvinotoolkit.org/latest/_docs_MO_DG_prepare_model_convert_model_Convert_Model_From_TensorFlow.html).


> [https://github.com/intel-analytics/analytics-zoo/blob/master/apps/model-inference-examples/model-inference-flink/src/main/scala/com/intel/analytics/zoo/apps/model/inference/flink/Resnet50ImageClassification/ImageClassificationStreaming.scala](url)
> 
> I have three questions hoped to be solved.
> Fisrt, from the name of the function, I thought the function "matToNCHWAndArray" is to
> change the BGR,HWC opencvmat to BGR,CWH.
> But we can see it wiil be given to a Jtensor of shape (1, 224,224,3) in the ImageClassificationStreaming.scala. I can't understand. Why not (1, 3, 244, 244)?
> Dose it mean the function "matToNCHWAndArray" didn't change the opencvmat to CHWArray?
> 
> Second, what's the preprocess for images of the Resnet50 model in the zoo?
> Is it just load the images ,resize to (244,244), change into np.float?
> not normalize? not divided by mean?
> 
> Third, why should I provide "inputShape: Array[Int], ifReverseInputChannels: Boolean, meanValues: Array[Float], scale: Float" such parameters for the function "loadtf()"?
> Does it will do such operation reverse the channels, divide the meanValus for the images I send to predict?
> 
> Here are some main code I can't understand from the example:
> 
> ```
> def matToNCHWAndArray(mat: OpenCVMat) = {
>     val (height, width, channel) = (mat.height(), mat.width(), mat.channels())
>     val data = new Array[Float](height * width * channel)
>     OpenCVMat.toFloatPixels(mat, data)
>     val resArray = new Array[Float](height * width * channel)
>     for (h <- 0 to height - 1) {
>       for (w <- 0 to width - 1) {
>         for (c <- 0 to 2) {
>           resArray(c * 224 * 224 + h * 224 + w) = data(h * 224 * 3 + w * 3 + c)
>         }
>       }
>     }
>     resArray
>   }
> ```
> 
> ImageClassificationStreaming.scala
> 
> ```
> val inputs = fileList.map(file => {
>       val imageBytes = FileUtils.readFileToByteArray(file)
>       val imageProcess = new ImageProcessor
>       val res = imageProcess.preProcess(imageBytes, 224, 224, 123, 116, 103, 1.0)
>       val input = new JTensor(res, Array(1, 224, 224, 3))
>       List(util.Arrays.asList(input)).asJava
>     })
> ```

**Comment by linjiaqin** [comment]
@qiyuangong 
Thank you!
Dose it mean that we needn't to do the preprocess just set the parameters ifReverseInputChannels = true, normalization and mean. OpenVino will do these things for us when we load a model by openvino to do a inference? 
For example, when I send a original image with size of (224,224) and channel of RGB to do inference, 
If I set "ifReverseInputChannels = true", the channel of the image will stay RGB?
If I set  the parameters normalization and mean, the openvino will help us automatically to change the image?  We needn't do such operators to preprocess when we do inference ?

**Comment by qiyuangong** [comment]
@linjiaqin 

First,  if`"ifReverseInputChannels` is set to be `true`. Then, input `RGB` will be reverse to `BGR`. So, If your input is RGB and model accept RGB, pls set it to be `false`.

Then, `normalization` (means and scale) in preprocessing.  Sometimes they are necessary for inference.  For example, `resnet_50` requires `normalization` in inference, otherwise accuracy will be worse. OpenVINO can do normalization for you when you do inference. 

In near future, OpenVINO can do more than just normalization, it will include resize and crop to further improve performance.

**Comment by qiyuangong** [comment]
Issue closed. All related discussion will be added to document if necessary. :)
